lib/src/multipart_request.dart

```diff
@@ -4,162 +4,143 @@
+      header.addAll(UTF8.encode(_headerForFile(file)));
+      contentLength += header.length + file.length + 2;
+      fileHeaders[i] = header;
+    }
+
+    // Ending characters
+    var ending = '--$boundary--\r\n'.codeUnits;
+    contentLength += ending.length;
+
+    // Write the files to the stream
+    //
+    // Future.forEach will ensure that the actions happen in sequence so i is
+    // used to get the fileHeaders
+    var i = 0;
+
+    Future.forEach(files, (file) {
+      assert(files[i] == file);
+      controller.add(fileHeaders[i++]);
+      return writeStreamToSink(file.read(), controller).then((_) => controller.add([13, 10]));
     }).then((_) {
       // TODO(nweiz): pass any errors propagated through this future on to
       // the stream. See issue 3657.
-      writeAscii('--$boundary--\r\n');
+      controller.add(ending);
       controller.close();
     });
 
-    return new ByteStream(controller.stream);
-  }
-
-  /// Returns the header string for a field. The return value is guaranteed to
-  /// contain only ASCII characters.
-  String _headerForField(String name, String value) {
-    var header =
-        'content-disposition: form-data; name="${_browserEncode(name)}"';
-    if (!isPlainAscii(value)) {
-      header = '$header\r\n'
-          'content-type: text/plain; charset=utf-8\r\n'
-          'content-transfer-encoding: binary';
-    }
-    return '$header\r\n\r\n';
+    return new MultipartBody._(controller.stream, contentLength);
   }
 
-  /// Returns the header string for a file. The return value is guaranteed to
-  /// contain only ASCII characters.
-  String _headerForFile(MultipartFile file) {
-    var header = 'content-type: ${file.contentType}\r\n'
-      'content-disposition: form-data; name="${_browserEncode(file.field)}"';
+  MultipartBody._(this._stream, this.contentLength);
 
-    if (file.filename != null) {
-      header = '$header; filename="${_browserEncode(file.filename)}"';
+  /// Returns a [Stream] representing the body.
+  ///
+  /// Can only be called once.
+  Stream<List<int>> read() {
+    if (_stream == null) {
+      throw new StateError("The 'read' method can only be called once on a "
+          "http.Request/http.Response object.");
     }
-    return '$header\r\n\r\n';
+    var stream = _stream;
+    _stream = null;
+    return stream;
   }
+}
 
-  /// Encode [value] in the same way browsers do.
-  String _browserEncode(String value) {
-    // http://tools.ietf.org/html/rfc2388 mandates some complex encodings for
-    // field names and file names, but in practice user agents seem not to
-    // follow this at all. Instead, they URL-encode `\r`, `\n`, and `\r\n` as
-    // `\r\n`; URL-encode `"`; and do nothing else (even for `%` or non-ASCII
-    // characters). We follow their behavior.
-    return value.replaceAll(_newlineRegExp, "%0D%0A").replaceAll('"', "%22");
+/// Returns the header string for a field.
+String _headerForField(String name, String value) {
```

Comment:
I generally make private methods that are only used by one class members of that class, even if they don't access its state directly.

Reply:
These are accessed from a factory constructor so I put them below. Do you want them as static class methods?

Reply:
Yes.